dlc/src/main/scala/org/bitcoins/dlc/DLCMessage.scala

```diff
@@ -2,37 +2,401 @@ package org.bitcoins.dlc
+      outcomeValueMap.map(f)
+
+    def flatMap[B](
+        f: ((Sha256DigestBE, Satoshis)) => GenTraversableOnce[B]): Iterable[B] =
+      outcomeValueMap.flatMap(f)
+  }
+
+  object ContractInfo extends Factory[ContractInfo] {
+
+    private val sizeOfMapElement: Int = 40
+
+    override def fromBytes(bytes: ByteVector): ContractInfo = {
+      @tailrec
+      def loop(
+          remainingBytes: ByteVector,
+          accum: Vector[(Sha256DigestBE, Satoshis)]): Vector[
+        (Sha256DigestBE, Satoshis)] = {
+        if (remainingBytes.size < sizeOfMapElement) {
+          accum
+        } else {
+          val relevantBytes = remainingBytes.take(sizeOfMapElement)
+          val digest = Sha256DigestBE(relevantBytes.take(32))
+          val sats = Satoshis(relevantBytes.takeRight(8))
+          loop(remainingBytes.drop(sizeOfMapElement), accum :+ (digest, sats))
+        }
+      }
+      ContractInfo(loop(bytes, Vector.empty).toMap)
+    }
+  }
 
   case class DLCOffer(
-      contractInfo: Map[Sha256DigestBE, Satoshis],
+      contractInfo: ContractInfo,
       oracleInfo: OracleInfo,
       extPubKey: ExtPublicKey,
       totalCollateral: Satoshis,
       fundingInputs: Vector[(TransactionOutPoint, TransactionOutput)],
       changeAddress: Bech32Address,
       feeRate: SatoshisPerVirtualByte,
       timeouts: DLCTimeouts)
-      extends DLCMessage
+      extends DLCMessage {
+
+    def toJson: Value = {
+      val contractInfosJson =
+        contractInfo
+          .map(
+            info =>
+              mutable.LinkedHashMap("sha256" -> Str(info._1.hex),
+                                    "sats" -> Num(info._2.toLong)))
+
+      val fundingInputsJson =
+        fundingInputs
+          .map(
+            input =>
+              mutable.LinkedHashMap("outpoint" -> Str(input._1.hex),
+                                    "output" -> Str(input._2.hex)))
+
+      val timeoutsJson =
+        mutable.LinkedHashMap(
+          "penalty" -> Num(timeouts.penaltyTimeout),
+          "contractMaturity" -> Num(timeouts.contractMaturity.toUInt32.toLong),
+          "contractTimeout" -> Num(timeouts.contractTimeout.toUInt32.toLong)
+        )
+
+      Obj(
+        mutable.LinkedHashMap[String, Value](
+          "contractInfo" -> contractInfosJson,
+          "oracleInfo" -> Str(oracleInfo.hex),
+          "extPubKey" -> Str(extPubKey.hex),
+          "totalCollateral" -> Num(totalCollateral.toLong),
+          "fundingInputs" -> fundingInputsJson,
+          "changeAddress" -> Str(changeAddress.value),
+          "feeRate" -> Num(feeRate.toLong),
+          "timeouts" -> timeoutsJson
+        )
+      )
+    }
+
+    def toJsonStr: String = toJson.toString()
+  }
+
+  object DLCOffer {
+
+    def fromJson(js: Value): DLCOffer = {
+      val iter = js.obj.iterator
+
+      val emptyOffer =
+        DLCOffer(
+          ContractInfo(Map.empty[Sha256DigestBE, Satoshis]),
+          null,
```

Comment:
!! null  is evil !!, this needs to change also, can you do give these things named arguments, it's much easier to code review. `alt+enter`

Reply:
all other instances of null needed to be removed to.